File: scalechain-api/src/test/scala/io/scalechain/blockchain/api/domain/RpcParamsDeserializationSpec.scala
package io.scalechain.blockchain.api.domain

import org.scalatest._
import spray.json.DefaultJsonProtocol._
import spray.json._

class RpcParamsDeserializationSpec extends FlatSpec with BeforeAndAfterEach with Matchers {
  this: Suite =>

  override def beforeEach() {
    // set-up code
    //

    super.beforeEach()
  }

  override def afterEach() {
    super.afterEach()
    // tear-down code
    //
  }

  import RpcParamsJsonFormat._

  implicit val implicitJsonRpcRequest = jsonFormat4(RpcRequest.apply)

  "RpcParams" should "be able to leave list of JsValues as is, if we want" in {

    val jsonrpcValue = "1.0"
    val id = 1
    val method = "mymethod"
    val params = JsArray(
      // a simple string
      JsString("1"),
      // a simple number
      JsNumber(1.5),
      // a simple array
      JsArray(
        JsString("2"),
        JsNumber(3.5)
      ),
      // an object with an array field
      JsObject(
        "field" -> JsString("3"),
        "arr" -> JsArray(
          JsString("4")
        )
      ),
      // an array of objects, which may have an array of values
      JsArray(
        JsObject(
          "field" -> JsString("5"),
          "arr" -> JsArray(
            JsString("6")
          )
        ),
        JsObject(
          "field" -> JsNumber(4.5)
        )
      )
    )

    val jsObject = JsObject(
      "jsonrpc" -> JsString(jsonrpcValue),
      "id" -> JsNumber(id),
      "method" -> JsString(method),
      "params" -> params
      )

    val request = jsObject.convertTo[RpcRequest]

    request.jsonrpc shouldBe Some(jsonrpcValue)
    request.id shouldBe id
    request.method shouldBe method
    request.params.paramValues shouldBe params.elements.toList
  }

  "RpcParams" should "not throw a DeserializationException even though jsonrpc field is missing " in {
    val jsObject = JsObject(
      //"jsonrpc" -> JsString("1.0"),
      "id" -> JsNumber(1),
      "method" -> JsString("myMethod"),
      "params" -> JsArray( JsString("arg1") )
    )

    jsObject.convertTo[RpcRequest]
  }

  "RpcParams" should "throw DeserializationException if id field is missing " in {
    val jsObject = JsObject(
      "jsonrpc" -> JsString("1.0"),
      //"id" -> JsString("abc"),
      "method" -> JsString("myMethod"),
      "params" -> JsArray( JsString("arg1") )
    )

    the [spray.json.DeserializationException] thrownBy {
      jsObject.convertTo[RpcRequest]
    } should have message "Object is missing required member 'id'"
  }

  "RpcParams" should "throw DeserializationException if method field is missing " in {
    val jsObject = JsObject(
      "jsonrpc" -> JsString("1.0"),
      "id" -> JsNumber(1),
      //"method" -> JsString("myMethod"),
      "params" -> JsArray( JsString("arg1") )
    )

    the [spray.json.DeserializationException] thrownBy {
      jsObject.convertTo[RpcRequest]
    } should have message "Object is missing required member 'method'"
  }

  "RpcParams" should "throw DeserializationException if params field is missing " in {
    val jsObject = JsObject(
      "jsonrpc" -> JsString("1.0"),
      "id" -> JsNumber(1),
      "method" -> JsString("myMethod")
      //"params" -> JsArray( JsString("arg1") )
    )

    the [spray.json.DeserializationException] thrownBy {
      jsObject.convertTo[RpcRequest]
    } should have message "Object is missing required member 'params'"
  }

  "RpcParams" should "throw DeserializationException if params field is not an array " in {
    val jsObject = JsObject(
      "jsonrpc" -> JsString("1.0"),
      "id" -> JsNumber(1),
      "method" -> JsString("myMethod"),
      "params" -> JsString("arg1")
    )

    the [spray.json.DeserializationException] thrownBy {
      jsObject.convertTo[RpcRequest]
    } should have message "JsArray expected for the params field."
  }

}

File: scalechain-api/src/test/kotlin/io/scalechain/blockchain/api/domain/RpcParamsDeserializationSpec.kt
package io.scalechain.blockchain.api.domain

import com.google.gson.*
import io.kotlintest.KTestJUnitRunner
import io.kotlintest.matchers.Matchers
import io.kotlintest.specs.FlatSpec
import io.scalechain.blockchain.api.Json
import org.junit.runner.RunWith

@RunWith(KTestJUnitRunner::class)
class RpcParamsDeserializationSpec : FlatSpec(), Matchers {

  override fun beforeEach() {
    // set-up code
    //

    super.beforeEach()
  }

  override fun afterEach() {
    super.afterEach()
    // tear-down code
    //
  }

  init {
    "RpcParams" should "be able to leave list of JsValues as is, if we want" {

      val jsonrpcValue = "1.0"
      val id = 1L
      val method = "mymethod"
      val arrayParam = JsonArray()
      arrayParam.add( JsonPrimitive("2") )
      arrayParam.add( JsonPrimitive(3.5) )

      val arrayParamInObject = JsonArray()
      arrayParamInObject.add(JsonPrimitive("4"))

      val objectParam = JsonObject()
      objectParam.add("field", JsonPrimitive("3"))
      objectParam.add("arr", arrayParamInObject)

      val objectArray = JsonArray()
      val obj1 = JsonObject()
      obj1.add("field", JsonPrimitive("5"))
      val obj1Arr = JsonArray()
      obj1Arr.add( JsonPrimitive("6"))
      obj1.add("arr", obj1Arr)
      val obj2 = JsonObject()
      obj2.add("field", JsonPrimitive(4.5))

      val params = JsonArray()
      // a simple string
      params.add(JsonPrimitive("1"))
      // a simple number
      params.add(JsonPrimitive(1.5))
      // a simple array
      params.add(arrayParam)
      // an object with an array field
      params.add(objectParam)
      // an array of objects, which may have an array of values
      params.add(objectArray)

      val jsObject = JsonObject()
      jsObject.add("jsonrpc", JsonPrimitive(jsonrpcValue) )
      jsObject.add("id", JsonPrimitive(id))
      jsObject.add("method", JsonPrimitive(method))
      jsObject.add("params", params )

      val request = Json.get().fromJson(jsObject, RpcRequest::class.java)

      request.jsonrpc shouldBe jsonrpcValue
      request.id shouldBe id
      request.method shouldBe method
      request.params.paramValues shouldBe params.toList()
    }

    "RpcParams" should "not throw a DeserializationException even though jsonrpc field is missing " {
      val jsObject = JsonObject()
      jsObject.add("id", JsonPrimitive(1))
      jsObject.add("method", JsonPrimitive("myMethod"))
      val params = JsonArray()
      params.add(JsonPrimitive("arg1"))
      jsObject.add("params", params)

      Json.get().fromJson(jsObject, RpcRequest::class.java)
    }

    "RpcParams" should "throw DeserializationException if id field is missing " {


      val jsObject = JsonObject()
      jsObject.add("jsonrpc", JsonPrimitive("1.0") )
      jsObject.add("method", JsonPrimitive("myMethod"))
      val params = JsonArray()
      params.add(JsonPrimitive("arg1"))
      jsObject.add("params", params)

      val thrown = shouldThrow <JsonSyntaxException> {
        Json.get().fromJson(jsObject, RpcRequest::class.java)
      }
      thrown.message shouldBe "Object is missing required member 'id'"
    }

    "RpcParams" should "throw DeserializationException if method field is missing " {

      val jsObject = JsonObject()
      jsObject.add("jsonrpc", JsonPrimitive("1.0") )
      jsObject.add("id", JsonPrimitive(1))
      val params = JsonArray()
      params.add(JsonPrimitive("arg1"))
      jsObject.add("params", params)

      val thrown = shouldThrow <JsonSyntaxException> {
        Json.get().fromJson(jsObject, RpcRequest::class.java)
      }
      thrown.message shouldBe "Object is missing required member 'method'"
    }

    "RpcParams" should "throw DeserializationException if params field is missing " {
      val jsObject = JsonObject()
      jsObject.add("jsonrpc", JsonPrimitive("1.0") )
      jsObject.add("id", JsonPrimitive(1))
      jsObject.add("method", JsonPrimitive("myMethod"))

      val thrown = shouldThrow <JsonSyntaxException> {
        Json.get().fromJson(jsObject, RpcRequest::class.java)
      }
      thrown.message shouldBe "Object is missing required member 'params'"
    }

    "RpcParams" should "throw DeserializationException if params field is not an array " {
      val jsObject = JsonObject()
      jsObject.add("jsonrpc", JsonPrimitive("1.0") )
      jsObject.add("id", JsonPrimitive(1))
      jsObject.add("method", JsonPrimitive("myMethod"))
      jsObject.add("params", JsonPrimitive("arg1"))

      val thrown = shouldThrow <JsonSyntaxException> {
        Json.get().fromJson(jsObject, RpcRequest::class.java)
      }
      thrown.message shouldBe "JsArray expected for the params field."
    }

  }

}
